### PR TITLE
CLOUDP-268810: Independent Custom Roles: Fixed endless reconciliations

### PR DESCRIPTION
**Pull request by igor-karpukhin**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

**Comment by s-urbaniak** [comment]
@igor-karpukhin jira ticket please :-)

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
**blocking**: I am missing a small unit test asserting the failure, being red before this fix and going green with this fix.

File: pkg/controller/atlascustomrole/atlascustomrole_controller.go [hunk 1]
```diff
@@ -165,8 +165,7 @@ func selectCredentials(ctx context.Context, k8sClient client.Client, akoRole *ak
 		}
 		// otherwise, use those attached to the AtlasProject that is referenced by the externalProjectRef
 		project := &akov2.AtlasProject{}
-		err := k8sClient.Get(ctx,
-			client.ObjectKey{Name: akoRole.Spec.ProjectRef.Name, Namespace: akoRole.Spec.ProjectRef.Namespace}, project)
+		err := k8sClient.Get(ctx, *akoRole.Spec.ProjectRef.GetObject(akoRole.Namespace), project)
```
Comment (s-urbaniak) [inline pkg/controller/atlascustomrole/atlascustomrole_controller.go:168]: is this related to the fix?

File: pkg/controller/atlascustomrole/custom_role.go [hunk 2]
```diff
@@ -37,7 +37,9 @@ func handleCustomRole(ctx *workflow.Context, k8sClient client.Client, akoCustomR
 		k8sClient: k8sClient,
 	}
 
-	return r.Reconcile()
+	result := r.Reconcile()
+	ctx.SetConditionFromResult(api.ReadyType, result)
```
Comment (s-urbaniak) [inline pkg/controller/atlascustomrole/custom_role.go:41]: this also looks like something unrelated to the fix?

File: pkg/controller/atlascustomrole/atlascustomrole_controller.go [hunk 3]
```diff
@@ -225,7 +224,7 @@ func (r *AtlasCustomRoleReconciler) notFound(req ctrl.Request) ctrl.Result {
 func (r *AtlasCustomRoleReconciler) SetupWithManager(mgr ctrl.Manager, skipNameValidation bool) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		Named("AtlasCustomRole").
-		For(&akov2.AtlasCustomRole{}).
+		For(&akov2.AtlasCustomRole{}, builder.WithPredicates(r.GlobalPredicates...)).
```
Comment (s-urbaniak) [inline pkg/controller/atlascustomrole/atlascustomrole_controller.go:227]: oh that was it, in that case, this LGTM.

**Review by s-urbaniak — APPROVED** [review]
Given the fix (adding global predicates) this LGTM and doesn't need a unit test 👍

**Review by josvazg — COMMENTED** [review]
Sorry but I do not understand the fix or the bug. Could we add a short explanation in the PR description?

**Review by helderjs — APPROVED** [review]

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
ooops, shouldn't have approved, this is missing a ticket, also as @josvazg mentioned, I'd love to see some explanation in the PR description.

File: pkg/controller/atlasproject/custom_roles.go [hunk 4]
```diff
@@ -31,12 +31,25 @@ func hasSkippedCustomRoles(atlasProject *akov2.AtlasProject) (bool, error) {
...
+		return len(lastSkippedSpec.CustomRoles) != 0, nil
 	}
 
 	return false, nil
 }
 
+func hasLastAppliedCustomRoles(atlasProject *akov2.AtlasProject) (bool, error) {
+	lastAppliedSpec := akov2.AtlasProjectSpec{}
+	lastAppliedSpecStr, ok := atlasProject.Annotations[customresource.AnnotationLastAppliedConfiguration]
+	if ok {
```
Comment (s-urbaniak) [inline pkg/controller/atlasproject/custom_roles.go:43]: **nit**: better readability
```suggestion
	if !ok {
		return false, nil
	}

	if err := json.Unmarshal([]byte(lastAppliedSpecStr), &lastAppliedSpec); err != nil {
		return false, fmt.Errorf("failed to parse last applied configuration: %w", err)
	}

	return len(lastAppliedSpec.CustomRoles) != 0, nil
```

**Review by s-urbaniak — APPROVED** [review]

File: pkg/controller/atlasproject/custom_roles.go [hunk 5]
```diff
@@ -31,12 +31,26 @@ func hasSkippedCustomRoles(atlasProject *akov2.AtlasProject) (bool, error) {
...
+	lastAppliedSpecStr, ok := atlasProject.Annotations[customresource.AnnotationLastAppliedConfiguration]
+	if !ok {
+		return false, nil
+	}
+
+	if err := json.Unmarshal([]byte(lastAppliedSpecStr), &lastAppliedSpec); err != nil {
+		return false, fmt.Errorf("failed to parse last applied configuration: %w", err)
+	}
+
+	return len(lastAppliedSpec.CustomRoles) != 0, nil
```
Comment (helderjs) [inline pkg/controller/atlasproject/custom_roles.go:51]: As many other sub-resources, it is a slice of items. This implementation is an all or nothing rule, as we allowed mixed modes, this doesn't allow different subsets of custom roles of being managed here and in a separate resource. Use case: Customer has a project with custom roles and start creating new ones from the new resource while plan the migration.
cc @igor-karpukhin @s-urbaniak

Comment (josvazg) [reply pkg/controller/atlasproject/custom_roles.go:51]: Sorry, I did not get that. Do you mean this change broke the migration process for custom roles?
